katcp/resource.py

```diff
@@ -690,6 +669,72 @@ def get_status(self):
         # By now the sensor manager should have set the reading
         raise Return(self._reading.status)
 
+    def wait(self, condition_or_value, status=None, timeout=None):
+        """Wait for sensor to satisfy a condition.
+
+        Parameters
+        ----------
+        condition_or_value : value, object or callable or seq
+            Conditions to be evaluated TODO, use SensorTranstionWaiter thingum
+        status : int enum, key of katcp.Sensor.SENSOR_TYPES or None
+            Wait for this status, at the same time as value above, to be
+            obtained. Ignore status if None
+        timeout : float or None
+            The timeout in seconds
+
+        Returns
+        -------
+        This command returns a tornado Future that resolves with True if the sensor values
+        satisifed the condition with the timeout, else resolves with False.
+
+        Raises
+        ------
+        Resolves with a :class:`KATCPSensorError` exception if the sensors does not
+        have a strategy set.
+
+        Resolves with :class:`tornado.gen.TimeoutError` if timeout is not None
+        and the sensor does not attain the requested value within the timeout.
+
+        """
+
+        if callable(condition_or_value) or (
+                isinstance(condition_or_value, collections.Sequence) and not
+                isinstance(condition_or_value, basestring)):
+            raise NotImplementedError(
+                'Currently only simple value-waits are supported')
+
+        ioloop = tornado.ioloop.IOLoop.current()
+        f = Future()
+        if self.sampling_strategy == ('none', ):
+            raise KATCPSensorError(
+                'Cannot wait on a sensor that does not have a strategy set')
+
+        def handle_update(sensor, reading):
+            try:
+                assert sensor is self
+                val_matched = reading.value == condition_or_value
+                status_matched = reading.status == status or status is None
+                if val_matched and status_matched:
+                    self.unregister_listener(handle_update)
+                    ioloop.add_callback(f.set_result, True)
+            except Exception:
+                f.set_exc_info(sys.exc_info())
+                self.unregister_listener(handle_update)
+
+        self.register_listener(handle_update)
+        ioloop.add_callback(handle_update, self, self._reading)
```

Comment:
Question? The ioloop.add_callback line will result in 'handle_update' being called on the next ioloop iteration ... so it is not to say that the sensor had an updated value yet. Where is the actual delay to allow the conditions to be met within 'timeout' time?